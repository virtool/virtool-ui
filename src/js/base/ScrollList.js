import React from "react";
import { map } from "lodash-es";
import { LoadingPlaceholder } from "./index";
import styled from "styled-components";

export const getScrollRatio = () =>
    ((window.innerHeight + window.scrollY) / document.documentElement.scrollHeight).toFixed(1);

const StyledScrollList = styled.div`
    padding-bottom: 20px;
    position: relative;
    z-index: 0;
`;

export class ScrollList extends React.Component {
    componentDidMount() {
        window.addEventListener("scroll", this.onScroll);
        this.prevPage = 0;
    }

    componentWillUnmount() {
        window.removeEventListener("scroll", this.onScroll);
    }

    onScroll = () => {
        if (
            this.props.page !== this.prevPage &&
            this.props.documents.length &&
            this.props.page < this.props.pageCount &&
            getScrollRatio() > 0.8
        ) {
            this.prevPage = this.props.page;
            this.props.onLoadNextPage(this.props.page + 1);
        }
    };

    render() {
        const { documents, renderRow, page, pageCount } = this.props;

        const entries = map(documents, (item, index) => renderRow(index));

        let loading;

        if (documents === null && page < pageCount) {
            loading = <LoadingPlaceholder margin="20px" />;
        }
        return (
            <StyledScrollList>
                {entries}
                {loading}
            </StyledScrollList>
        );
    }
}
