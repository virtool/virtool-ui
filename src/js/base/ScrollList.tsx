import { FetchNextPageOptions, InfiniteQueryObserverResult } from "@tanstack/react-query";
import { map } from "lodash-es";
import React, { useEffect } from "react";
import styled from "styled-components";
import { LoadingPlaceholder } from "./LoadingPlaceholder";

function getScrollRatio(): number {
    return (window.innerHeight + window.scrollY) / document.documentElement.scrollHeight;
}

const StyledScrollList = styled.div`
    margin-bottom: 20px;
    position: relative;
    z-index: 0;
`;

type ScrollListProps = {
    /** The class name of the scroll list */
    className?: string;
    /** A function which initiates fetching the next page */
    fetchNextPage: (options?: FetchNextPageOptions) => Promise<InfiniteQueryObserverResult>;
    /** Whether there is another page of data available */
    hasNextPage: boolean;
    /** Whether a new page is being fetched */
    isFetchingNextPage: boolean;
    /** Whether the first page is being fetched */
    isLoading: boolean;
    /** The list of items */
    items: unknown[];
    /** A function which accepts an item and returns a React element */
    renderRow: (item: unknown) => void;
};

/**
 * An infinitely scrolling list of items.
 */
export function ScrollList({
    className,
    fetchNextPage,
    isFetchingNextPage,
    hasNextPage,
    isLoading,
    items,
    renderRow,
}: ScrollListProps) {
    useEffect(() => {
        handleFetchNextPage();

        function handleFetchNextPage() {
            if (getScrollRatio() > 0.8 && !isFetchingNextPage && hasNextPage) {
                void fetchNextPage();
            }
        }

        window.addEventListener("scroll", handleFetchNextPage);
        return () => window.removeEventListener("scroll", handleFetchNextPage);
    }, [isFetchingNextPage, fetchNextPage]);

    const entries = map(items, item => renderRow(item));

    return (
        <StyledScrollList className={className}>
            {entries}
            {isLoading && <LoadingPlaceholder margin="20px" />}
        </StyledScrollList>
    );
}
