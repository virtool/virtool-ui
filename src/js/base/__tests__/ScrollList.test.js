import { ScrollList } from "../ScrollList";

describe("<ScrollList />", () => {
    let props;

    beforeEach(() => {
        props = {
            isElement: true,
            document: [{ foo: "bar" }],
            page: 1,
            pageCount: 2,
            onLoadNextPage: jest.fn()
        };
    });

    it("should return LoadingPlaceholder when [documents=null] and [page<pageCount]", () => {
        props.documents = null;
        const wrapper = mount(<ScrollList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should return React.Fragment when [noContainer=true]", () => {
        const wrapper = mount(<ScrollList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [noContainer=false]", () => {
        props.noContainer = false;
        const wrapper = mount(<ScrollList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("componentDidMount() should call addEventListener()", () => {
        window.addEventListener = jest.fn();
        window.removeEventListener = jest.fn();

        const wrapper = mount(<ScrollList {...props} />);

        expect(window.addEventListener).toHaveBeenCalledWith("scroll", wrapper.instance().onScroll);
    });

    it("componentWillUnmount() should call removeEventListener()", () => {
        window.addEventListener = jest.fn();
        window.removeEventListener = jest.fn();

        const wrapper = mount(<ScrollList {...props} />);
        const onScroll = wrapper.instance().onScroll;
        wrapper.unmount();

        expect(window.removeEventListener).toHaveBeenCalledWith("scroll", onScroll);
    });
});
