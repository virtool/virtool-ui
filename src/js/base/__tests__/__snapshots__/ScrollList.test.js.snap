// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollList /> should render when [noContainer=false] 1`] = `
.c0 {
  padding-bottom: 20px;
  position: relative;
  z-index: 0;
}

<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  isElement={true}
  noContainer={false}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <ScrollList__StyledScrollList>
    <div
      className="c0"
    />
  </ScrollList__StyledScrollList>
</ScrollList>
`;

exports[`<ScrollList /> should return LoadingPlaceholder when [documents=null] and [page<pageCount] 1`] = `
.c2 {
  -webkit-animation: dzGqrz 0.75s 0s infinite linear;
  animation: dzGqrz 0.75s 0s infinite linear;
  border: 2px solid;
  border-bottom-color: transparent !important;
  border-radius: 100%;
  background: transparent;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  display: inline-block;
  height: 22px;
  width: 22px;
}

.c1 {
  margin-top: 20px;
  text-align: center;
}

.c0 {
  padding-bottom: 20px;
  position: relative;
  z-index: 0;
}

<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  documents={null}
  isElement={true}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <ScrollList__StyledScrollList>
    <div
      className="c0"
    >
      <LoadingPlaceholder
        margin="20px"
      >
        <LoadingPlaceholder__StyledLoadingPlaceholder
          margin="20px"
        >
          <div
            className="c1"
          >
            <Loader>
              <Loader__StyledLoader
                color="greyDark"
                size="22px"
              >
                <div
                  className="c2"
                  color="greyDark"
                  size="22px"
                >
                  <div />
                </div>
              </Loader__StyledLoader>
            </Loader>
          </div>
        </LoadingPlaceholder__StyledLoadingPlaceholder>
      </LoadingPlaceholder>
    </div>
  </ScrollList__StyledScrollList>
</ScrollList>
`;

exports[`<ScrollList /> should return React.Fragment when [noContainer=true] 1`] = `
.c0 {
  padding-bottom: 20px;
  position: relative;
  z-index: 0;
}

<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  isElement={true}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <ScrollList__StyledScrollList>
    <div
      className="c0"
    />
  </ScrollList__StyledScrollList>
</ScrollList>
`;
